File: app/gululu_comment_service.py
"""Gululu online-comment cache with explicit stale and failure results."""
from __future__ import annotations

import logging
import time
from pathlib import Path
from typing import Callable

from .gululu_client import GululuClient
from .gululu_comments import comment_to_public
from .gululu_source import parse_book_id
from .storage import atomic_write_json, load_json_file, now_iso


log = logging.getLogger("gululu_service")
_COMMENT_CACHE_TTL_SECONDS = 300
_MAX_COMMENT_SCOPES = 64


class GululuCommentService:
    def __init__(
        self,
        library_dir: Callable[[], Path],
        client_factory: Callable[[], GululuClient],
    ) -> None:
        self._library_dir = library_dir
        self._client_factory = client_factory

    def get_comments(
        self,
        source: str | int,
        floor_ids: list[int],
        *,
        refresh: bool = False,
    ) -> dict:
        try:
            source_id = parse_book_id(source)
            scopes = self._validate_scopes(floor_ids)
        except ValueError as exc:
            return {"ok": False, "error": str(exc), "floors": []}

        cached = {floor_id: self._read_cache(source_id, floor_id) for floor_id in scopes}
        pending = [
            floor_id
            for floor_id in scopes
            if refresh or cached[floor_id] is None or not cached[floor_id]["fresh"]
        ]
        network_error = ""
        if pending:
            try:
                with self._client_factory() as client:
                    fetched = client.fetch_comments(source_id, pending)
                for floor_id in pending:
                    comments = [comment_to_public(item) for item in fetched.get(floor_id, [])]
                    payload = {
                        "version": 1,
                        "source_id": source_id,
                        "floor_id": floor_id,
                        "fetched_at": now_iso(),
                        "comments": comments,
                    }
                    self._write_cache(source_id, floor_id, payload)
                    cached[floor_id] = {"payload": payload, "fresh": True}
            except Exception as exc:  # noqa: BLE001 - converted to explicit API results below
                network_error = str(exc)
                log.warning(
                    "骨碌碌在线评论读取失败 source_id=%s floors=%s: %s",
                    source_id,
                    pending,
                    exc,
                )

        floors = []
        hard_errors = []
        for floor_id in scopes:
            item = cached.get(floor_id)
            if item is None:
                error = network_error or "评论缓存不可用"
                hard_errors.append(f"{floor_id}: {error}")
                floors.append({
                    "floor_id": floor_id,
                    "comments": [],
                    "cached": False,
                    "stale": False,
                    "fetched_at": "",
                    "error": error,
                })
                continue
            payload = item["payload"]
            was_pending = floor_id in pending
            stale = bool(network_error and was_pending)
            floors.append({
                "floor_id": floor_id,
                "comments": payload["comments"],
                "cached": not was_pending or stale,
                "stale": stale,
                "fetched_at": payload["fetched_at"],
                "error": network_error if stale else "",
            })
        return {
            "ok": not hard_errors,
            "source_id": source_id,
            "floors": floors,
            "error": "; ".join(hard_errors),
        }

    @staticmethod
    def _validate_scopes(floor_ids: list[int]) -> list[int]:
        if not isinstance(floor_ids, list):
            raise ValueError("评论楼层列表格式错误")
        scopes = list(dict.fromkeys(floor_ids))
        if not scopes or len(scopes) > _MAX_COMMENT_SCOPES:
            raise ValueError(f"单次评论请求必须包含 1-{_MAX_COMMENT_SCOPES} 个楼层")
        if any(isinstance(value, bool) or not isinstance(value, int) or value < 0 for value in scopes):
            raise ValueError("评论楼层 ID 格式错误")
        return scopes

    def _cache_path(self, source_id: int, floor_id: int) -> Path:
        return self._library_dir() / str(source_id) / "comments" / f"{floor_id}.json"

    def _read_cache(self, source_id: int, floor_id: int):
        path = self._cache_path(source_id, floor_id)
        payload = load_json_file(path)
        if not isinstance(payload, dict):
            return None
        if (
            payload.get("version") != 1
            or payload.get("source_id") != source_id
            or payload.get("floor_id") != floor_id
            or not isinstance(payload.get("fetched_at"), str)
            or not isinstance(payload.get("comments"), list)
        ):
            return None
        try:
            fresh = time.time() - path.stat().st_mtime <= _COMMENT_CACHE_TTL_SECONDS
        except OSError:
            fresh = False
        return {"payload": payload, "fresh": fresh}

    def _write_cache(self, source_id: int, floor_id: int, payload: dict) -> None:
        path = self._cache_path(source_id, floor_id)
        path.parent.mkdir(parents=True, exist_ok=True)
        atomic_write_json(path, payload)

File: android/app/src/main/java/io/github/gighi947/ankeshelf/service/GululuCommentService.kt
package io.github.gighi947.ankeshelf.service

import io.github.gighi947.ankeshelf.data.AppPaths
import io.github.gighi947.ankeshelf.data.GululuComments
import io.github.gighi947.ankeshelf.data.atomicWriteJson
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.put
import java.io.File

/** 单个评论作用域的读取结果（cached/stale/error 显式区分，便于界面提示）。 */
data class GululuCommentScope(
    val floorId: Int,
    val comments: List<JsonObject>,
    val cached: Boolean,
    val stale: Boolean,
    val fetchedAt: String,
    val error: String,
)

data class GululuCommentResult(
    val ok: Boolean,
    val sourceId: Int,
    val floors: List<GululuCommentScope>,
    val error: String,
)

/**
 * 骨碌碌在线评论缓存（Kotlin 版 `app/gululu_comment_service.py`）。
 *
 * 语义与桌面一致：
 * - 每个作用域一份端私有缓存 `gululu_library/<id>/comments/<floorId>.json`（不入双端契约）；
 * - 5 分钟内视为新鲜，过期或强制刷新才走网络；
 * - **网络失败不抛给界面**：有缓存则回退为 `stale=true` 并带上错误原因，
 *   完全没有缓存才算硬失败（`ok=false`）——离线可读是这条链路的核心不变量。
 */
class GululuCommentService(
    private val appPaths: AppPaths,
    /** 注入点：默认走真实客户端；测试与离线场景传入假实现（不需要打开 client 类）。 */
    private val fetchComments: (bookId: Int, floorIds: List<Int>) -> Map<Int, List<JsonObject>> =
        { bookId, floorIds -> GululuClient().fetchComments(bookId, floorIds) },
    private val nowMillis: () -> Long = { System.currentTimeMillis() },
) {
    private val json = Json { ignoreUnknownKeys = true; prettyPrint = true }

    fun getComments(
        sourceId: Int,
        floorIds: List<Int>,
        refresh: Boolean = false,
    ): GululuCommentResult {
        val scopes = floorIds.distinct()
        if (scopes.isEmpty() || scopes.size > MAX_COMMENT_SCOPES) {
            return GululuCommentResult(
                ok = false,
                sourceId = sourceId,
                floors = emptyList(),
                error = "单次评论请求必须包含 1-$MAX_COMMENT_SCOPES 个楼层",
            )
        }
        if (scopes.any { it < 0 }) {
            return GululuCommentResult(false, sourceId, emptyList(), "评论楼层 ID 格式错误")
        }

        val cached = scopes.associateWith { readCache(sourceId, it) }.toMutableMap()
        val pending = scopes.filter { refresh || cached[it]?.fresh != true }
        var networkError = ""
        if (pending.isNotEmpty()) {
            try {
                val fetched = fetchComments(sourceId, pending)
                for (floorId in pending) {
                    val comments = (fetched[floorId] ?: emptyList()).map {
                        GululuComments.commentToPublic(it)
                    }
                    val payload = buildJsonObject {
                        put("version", 1)
                        put("source_id", sourceId)
                        put("floor_id", floorId)
                        put("fetched_at", io.github.gighi947.ankeshelf.data.nowIso())
                        put("comments", JsonArray(comments))
                    }
                    writeCache(sourceId, floorId, payload)
                    cached[floorId] = CachedScope(payload, fresh = true)
                }
            } catch (e: Exception) {
                // 显式降级：记录原因并回退缓存，绝不把网络抖动变成"评论功能不可用"。
                networkError = e.message?.takeIf { it.isNotBlank() } ?: e.javaClass.simpleName
                LogEvents.event(
                    "gululu",
                    "comments_fetch_failed",
                    "source_id" to sourceId,
                    "floors" to pending.size,
                    "error" to networkError,
                )
            }
        }

        val floors = mutableListOf<GululuCommentScope>()
        val hardErrors = mutableListOf<String>()
        for (floorId in scopes) {
            val item = cached[floorId]
            if (item == null) {
                val error = networkError.ifEmpty { "评论缓存不可用" }
                hardErrors.add("$floorId: $error")
                floors.add(GululuCommentScope(floorId, emptyList(), false, false, "", error))
                continue
            }
            val wasPending = floorId in pending
            val stale = networkError.isNotEmpty() && wasPending
            floors.add(
                GululuCommentScope(
                    floorId = floorId,
                    comments = (item.payload["comments"] as? JsonArray)
                        ?.filterIsInstance<JsonObject>() ?: emptyList(),
                    cached = !wasPending || stale,
                    stale = stale,
                    fetchedAt = (item.payload["fetched_at"] as? JsonPrimitive)?.content.orEmpty(),
                    error = if (stale) networkError else "",
                ),
            )
        }
        return GululuCommentResult(
            ok = hardErrors.isEmpty(),
            sourceId = sourceId,
            floors = floors,
            error = hardErrors.joinToString("; "),
        )
    }

    private data class CachedScope(val payload: JsonObject, val fresh: Boolean)

    private fun cacheFile(sourceId: Int, floorId: Int): File =
        File(appPaths.gululuLibraryDir, "$sourceId/comments/$floorId.json")

    private fun readCache(sourceId: Int, floorId: Int): CachedScope? {
        val file = cacheFile(sourceId, floorId)
        if (!file.isFile) return null
        val payload = runCatching { json.parseToJsonElement(file.readText()) as? JsonObject }
            .getOrNull() ?: return null
        val versionOk = (payload["version"] as? JsonPrimitive)?.intOrNull == 1
        val sourceOk = (payload["source_id"] as? JsonPrimitive)?.intOrNull == sourceId
        val floorOk = (payload["floor_id"] as? JsonPrimitive)?.intOrNull == floorId
        val fetchedAt = payload["fetched_at"] as? JsonPrimitive
        val comments = payload["comments"] as? JsonArray
        if (!versionOk || !sourceOk || !floorOk || fetchedAt == null || comments == null) return null
        val fresh = (nowMillis() - file.lastModified()) <= CACHE_TTL_MILLIS
        return CachedScope(payload, fresh)
    }

    private fun writeCache(sourceId: Int, floorId: Int, payload: JsonObject) {
        val file = cacheFile(sourceId, floorId)
        file.parentFile?.mkdirs()
        atomicWriteJson(file, json.encodeToString(JsonObject.serializer(), payload))
    }

    companion object {
        const val MAX_COMMENT_SCOPES = 64
        const val CACHE_TTL_MILLIS = 300_000L
    }
}
